src/Generation.ts

```diff
@@ -108,23 +109,41 @@ export default class Generation {
 		// If a thread for pc already exists in this generation, combine traces and return
 		const existingThreadForProgramCounter = this._threadsByProgramCounter[pc];
 		if (existingThreadForProgramCounter) {
-			// Detect repetition in the same generation, which would cause cyclic traces
-			if (!parentThread || !parentThread.trace.contains(pc, this._generationNumber)) {
-				// Non-cyclic trace, join threads
-				existingThreadForProgramCounter.join(parentThread, badness);
+			if (parentThread === null || parentThread === undefined) {
+				existingThreadForProgramCounter.updateBadness(badness);
+				return null;
+			}
+
+			if (parentThread.generationNumber !== this._generationNumber) {
+				existingThreadForProgramCounter.updateBadness(badness);
+				existingThreadForProgramCounter.trace.join(parentThread.trace);
+				return null;
 			}
 
+			const parentThreadRep = this._ancestries.find(parentThread.threadId);
+			if (parentThreadRep === existingThreadForProgramCounter.threadId) {
```

Comment:
This here will not correctly detect cycles with the following form:
```
Threads:
1 -> 2
2 -> 3
3 -> 2
```

We can not correctly detect these cycles. If we `find` both threads, we will be able to detect these two threads, but then we will also (incorrectly) detect the following case as a cycle:

```
Thread list:

1 -> 2, 3
2 -> 3
```

Reply:
Good catch! Let's close this PR for now, and revisit / reinvestigate this approach later if / when further optimization is required.